asdf/commands/edit.py

```diff
@@ -0,0 +1,589 @@
+
+    edited_text, delim = get_yaml_text_and_delimiter(edited_text)
+
+    padded_text = edited_text + b"\n" + b" " * (diff - 1) + delim
+    return padded_text, diff - 1
+
+
+def add_pad_to_new_text(edited_text, pad_size):
+    """
+    Adds pad to edited text.
+
+    Parameters
+    ----------
+    edited_text - The text from the edited YAML file.
+    pad_size - The number of spaces to add as a pad.
+
+    Return
+    ------
+    Pad text with the number of spaces requested as pad.
+    """
+
+    edited_text, delim = get_yaml_text_and_delimiter(edited_text)
+
+    pad = b" " * pad_size
+    padded_text = edited_text + b"\n" + pad + delim
+
+    return padded_text
+
+
+def write_block_index(fd, index):
+    """
+    Write the block index to an ASDF file.
+
+    Parameters
+    ----------
+    fd - The output file to write the block index.
+    index - A list of locations for each block.
+    """
+    if len(index) < 1:
+        return
+
+    # TODO - this needs to be changed to use constants.py and pyyaml
+    bindex_hdr = b"#ASDF BLOCK INDEX\n%YAML 1.1\n---\n"
+    fd.write(bindex_hdr)
+    for idx in index:
+        ostr = f"- {idx}\n"
+        fd.write(ostr.encode("utf-8"))
+    end = b"..."
+    fd.write(end)
+    return
+
+
+def get_next_block_header(fd):
+    """
+    From a file, gets the next block header.
+
+    Parameters
+    ----------
+    fd - The ASDF file to get the next block.
+
+    Return
+    ------
+    If a block is found, return the bytes of the block header.
+    Otherwise return None.
+    """
+    #     Block header structure:
+    # 4 bytes of magic number
+    # 2 bytes of header length, after the length field (min 48)
+    # 4 bytes flag
+    # 4 bytes compression
+    # 8 bytes allocated size
+    # 8 bytes used (on disk) size
+    # 8 bytes data size
+    # 16 bytes checksum
+    blk_header = fd.read(6)
+    if len(blk_header) != 6:
+        return None
+    if not blk_header.startswith(constants.BLOCK_MAGIC):
+        return None
+    hsz = struct.unpack(">H", blk_header[4:6])[0]
+    header = fd.read(hsz)
+    return blk_header + header
+
+
+def rewrite_asdf_file(edited_text, orig_text, oname, fname):
+    """
+    Rewrite an ASDF file for too large edited YAML.  The edited YAML, a pad,
+    the blocks will be rewritten.  A block index will also be rewritten.  If a
+    block index existed in the old file, it will have to be recomputed to
+    because of the larger YAML size and pad, which changes the location of
+    the binary blocks.
+
+    Parameters
+    ----------
+    edited_text : the new YAML text to write out.
+    orig_text : the original YAML text to overwrite.
+    oname : the ASDF file to overwrite.
+    fname : the edit YAML to write to new file.
+    """
+
+    tmp_oname = oname + ".tmp"  # Save as a temp file, in case anything goes wrong.
+    pad_size = 10 * 1000
+    padded_text = add_pad_to_new_text(edited_text, pad_size)
+
+    ifd = open(oname, "r+b")  # Open old ASDF to get binary blocks
```

Comment:
The `+` opens this file for reading _and_ writing, which I don't think we want here.

Reply:
I removed it.